Add a new method to PageView, called when the page changes and scrolling event ends.

tango5614 opened the pull request:
## Description
<b>What I have done?</b>

Add a new method called `pageDidChanged` to PageView which will be called at the end of the page change animation.

<b>Why did I do it?</b>
There's always situation where people want to do something after the page changes completely.
For example, in #28096 he/she wants to make the `pageView` unscrollable after it changes to a specific page. But that is impossible in `onPageChanged`, because it is called in the middle of page changing, which makes the `pageView` die.That's why we need a new method like `pageDidChange`.
## Related Issues

#28096

## Tests

I added the following tests:
Page changes and scroll animation ends
in the file `packages/flutter/lib/src/widgets/page_view.dart`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

tango5614 committed:
Add a new method to PageView which will be called at the end of the page change animation. (#28096)

tango5614 committed:
Merge branch 'master' into fix28096

* master: (25 commits)
  Increase daemon protocol version for getSupportedPlatforms (#33980)
  skip web test on crazy import (#34017)
  Compatibility pass on flutter/foundation tests for JavaScript compilation. (1) (#33349)
  0602dbb27 Roll src/third_party/dart 9dcb026b26..6e0d978505 (72 commits) (#34027)
  Add chrome stable to dockerfile and web shard (#33787)
  Codegen an entrypoint for flutter web applications (#33956)
  Revert "Reland "Text inline widgets, TextSpan rework" (#33946)" (#34002)
  Revert "Re-add deprecated method for plugin migration compatibility. (#34006)" (#34022)
  Remove print (#34004)
  Manual roll the engine to land the timing API (#33989)
  Make plugins Swift-first on macOS (#33997)
  Re-add deprecated method for plugin migration compatibility. (#34006)
  make sure version check includes hotfixes (#33459)
  Respond to AndroidView focus events. (#33901)
  Add 'doctor' support for Windows (#33872)
  Add use_frameworks to macOS Podfile template (#33987)
  [Material] Create a themable Range Slider (continuous and discrete) (#31681)
  Updating names to correct versioning convention (#33865)
  Whitelist adb.exe heap corruption exit code. (#33951)
  [flutter_tool] Fix 'q' for Fuchsia profile/debug mode (#33846)
  ...

tango5614 commented:
@goderbauer 
Thanks for your awesome framework.
Would you please check this PR out?

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -539,7 +542,8 @@ class PageView extends StatefulWidget {
 
   /// Called whenever the page in the center of the viewport changes.
   final ValueChanged<int> onPageChanged;
```
Can you update this documentation to include a sentence about how this is different from pageDidChanged?

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -539,7 +542,8 @@ class PageView extends StatefulWidget {
   /// Called whenever the page in the center of the viewport changes.
   final ValueChanged<int> onPageChanged;
-
+  /// Called whenever the page has changed and the scroll animation has stopped.
```
This doc should point out how it is different from onPageChanged.

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -556,12 +560,16 @@ class PageView extends StatefulWidget {
   void initState() {
     super.initState();
-    _lastReportedPage = widget.controller.initialPage;
+    /// a flag to decide whether to call [PageView.onPageChanged];
```
This comment should be on the property above.

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -556,12 +560,16 @@ class PageView extends StatefulWidget {
-    _lastReportedPage = widget.controller.initialPage;
+    /// a flag to decide whether to call [PageView.onPageChanged];
+    _onPageChangedIndex = widget.controller.initialPage;
+    /// a flag to decide whether to call [PageView.pageDidChanged];
```
same here

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -556,12 +560,16 @@ class PageView extends StatefulWidget {
 
 class _PageViewState extends State<PageView> {
-  int _lastReportedPage = 0;
+  int _onPageChangedIndex = 0;
```
Both of these names should probably still indicate that it is the last reported index.

goderbauer commented on packages/flutter/test/widgets/page_view_test.dart:
```diff
@@ -433,6 +461,7 @@ void main() {
         child: PageView(
           pageSnapping: pageSnapping,
           onPageChanged: log.add,
+          pageDidChanged: log.add,
```
There should probably be a different log variable so we can differentiate in the test which callback fired.

goderbauer commented on packages/flutter/test/widgets/page_view_test.dart:
```diff
@@ -580,13 +614,16 @@ void main() {
             changeIndex = page;
           },
+          pageDidChanged: (int page) {
+            changeIndex = page;
```
same here. This should set a different variable.

tango5614 committed:
Merge branch 'master' into fix28096

* master: (84 commits)
  Compatibility pass on flutter/material tests for JavaScript compilation. (9) (#33378)
  fix Applying decoration for a table row widget will cause render exception (#34285)
  Add widget of the week videos (#34356)
  Roll engine ab5c14b9491a..67aadb62068a (2 commits) (#34350)
  Re-apply compressionState changes. (#34341)
  Revert "update the Flutter.Frame event to use new engine APIs (#34243)" (#34352)
  Change Xcode project developmentRegion to 'en' and plist CFBundleDevelopmentRegion to DEVELOPMENT_LANGUAGE (#34293)
  update the Flutter.Frame event to use new engine APIs (#34243)
  [flutter_tool] Don't truncate verbose logs from _flutter.listViews (#34255)
  ab5c14b94 Set Dart version to git hash 3166bbf24b0c929eef33fd5d0f69e0f36a9009f3 (Dart 2.3.3-dev) (flutter/engine#9292) (#34336)
  Prepare for Uint8List SDK breaking changes (#34295)
  Roll engine b14d9719607a..d3c213e5cc63 (3 commits) (#34331)
  Roll engine 4fc95ebb294c..b14d9719607a (4 commits) (#34310)
  4fc95ebb2 Fixed memory leaks within FlutterFragment and FlutterView (#34268, #34269, #34270). (flutter/engine#9288) (#34305)
  Roll engine 2d2cfc0884ff..de350c4cbb4a (2 commits) (#34302)
  Roll engine 02e6a130e7dd..2d2cfc0884ff (3 commits) (#34292)
  Split gradle_plugin_test.dart (#34282)
  Roll engine 4d6847466d0b..02e6a130e7dd (4 commits) (#34281)
  Compatibility pass on flutter/widgets tests for JavaScript compilation. (8) (#33377)
  Instrument usage of include_flutter.groovy and xcode_backend.sh (#34189)
  ...

tango5614 committed:
Add a new method to PageView, called when the page changes and scrolling event ends. (#34039)

tango5614 commented:
Thanks for the instruction.All done @goderbauer

tango5614 committed:
Merge branch 'master' into fix28096

* master:
  More removing of timeouts. (#33932)
  Update flakiness of tests (#34374)
  Roll engine 67aadb62068a..209250da1bb3 (4 commits) (#34370)
  Don't show scrollbar if there isn't enough content (#34175)

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -619,8 +622,13 @@ class PageView extends StatefulWidget {
 
   /// Called whenever the page in the center of the viewport changes.
+  ///
+  /// It is called before [pageDidChanged], in the middle of scroll animation.
```
Isn't it called at the start of the transition?

tango5614 replied:
Actually it is called in the middle.According to the source code in `page_view.dart` line685-691
```
        if (notification.depth == 0 && widget.onPageChanged != null && notification is ScrollUpdateNotification) {
          final PageMetrics metrics = notification.metrics;
          final int currentPage = metrics.page.round(); 
          if (currentPage != _lastOnPageChangedIndex) {
            _lastOnPageChangedIndex = currentPage;
            widget.onPageChanged(currentPage);
          }
```
Only if the [metrics.page] is beyond 0.5, you can detect the page changing;
And this point is proved in the test script - `page_view_test.dart`
BTW, sorry for the grammatical mistakes.

goderbauer replied:
You're right! Thanks for checking.

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -619,8 +622,13 @@ class PageView extends StatefulWidget {
-
+  /// Called whenever the page has changed and the scroll animation has stopped.
+  ///
+  /// It is called after [onPageChanged], at the end of scroll animation.
```
nit: ... at the end of the scroll animation. (same above)

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -619,8 +622,13 @@ class PageView extends StatefulWidget {
+  /// Called whenever the page has changed and the scroll animation has stopped.
+  ///
+  /// It is called after [onPageChanged], at the end of scroll animation.
+  final ValueChanged<int> pageDidChanged;
```
Should this be "pageDidChange" to be grammatically correct?

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -637,12 +645,19 @@ class PageView extends StatefulWidget {
+
+  /// A flag to record last index reported by [PageView.pageDidChange] and to decide
+  /// whether to call [PageView.pageDidChanged];
+  int _lastPageDidChangedIndex = 0;
```
same here: DidChange

tango5614 committed:
Merge branch 'master' into fix28096

* master: (24 commits)
  [flutter_tool,fuchsia] Update the install flow for packaging migration. (#34447)
  SliverFillRemaining flag for different use cases (#33627)
  SizedBox documentation (#34424)
  Change API doc link to api.dart.dev (#34388)
  2589785b5 Roll src/third_party/skia 87e885038893..c3252a04b377 (3 commits) (flutter/engine#9327) (#34484)
  ace5d5907 Fix rawTypes errors in Android embedding classes (flutter/engine#9326) (#34481)
  bf0def61e Roll src/third_party/skia 4c4945a25248..87e885038893 (1 commits) (flutter/engine#9325) (#34471)
  Roll engine f1d821d8fde7..6f5347c5d6bf (13 commits) (#34466)
  Allow "from" hero state to survive hero animation in a push transition (#32842)
  Roll pub dependencies (#33677)
  Skip flaky test on Windows (#34464)
  Allow flaky tests to pass or fail and mark web tests as flaky (#34456)
  Dont depend on web SDK unless running tests on chrome (#34457)
  Fix semantics_tester (#34368)
  Include raw value in Diagnostics json for basic types (#34417)
  Refactor Gradle plugin (#34353)
  Allow web tests to fail in cirrus config (#34436)
  skip bottom_sheet (#34430)
  Remove unused flag `--target-platform` from `flutter run` (#34369)
  Extract DiagnosticsNode serializer from WidgetInspector (#34012)
  ...

tango5614 committed:
Add a new method to PageView, called when the page changes and scrolling event ends. (#34039) FIX grammatical mistakes.

Piinks commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -619,8 +622,13 @@ class PageView extends StatefulWidget {
+  /// It is called before [pageDidChange], in the middle of the scroll animation.
   final ValueChanged<int> onPageChanged;
-
+  /// Called whenever the page has changed and the scroll animation has stopped.
```
Nit: Please place a line break between each of these properties.
```suggestion

  /// Called whenever the page has changed and the scroll animation has stopped.
```

Piinks commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -619,8 +622,13 @@ class PageView extends StatefulWidget {
+  ///
+  /// It is called after [onPageChanged], at the end of the scroll animation.
+  final ValueChanged<int> pageDidChange;
   /// A delegate that provides the children for the [PageView].
```
```suggestion

  /// A delegate that provides the children for the [PageView].
```
line break

Piinks reviewed (commented):
Thanks for your work on this! There are just a few nits below.

tango5614 committed:
Add a new method to PageView, called when the page changes and scrolling event ends.Add a new method to PageView, called when the page changes and scrolling event ends. (#34039) Fix some nits.

tango5614 committed:
Merge branch 'master' into fix28096

* master:
  retry on HttpException (#34526)
  308c11301 Roll src/third_party/skia 057a32d9a2c8..5fe7429babe2 (1 commits) (flutter/engine#9342) (#34583)
  Add back ability to override the local engine in Gradle (#34460)
  Added customizable padding for the segmented controll (#34555)
  466a1d848 Roll src/third_party/skia 3721688f64a5..057a32d9a2c8 (1 commits) (flutter/engine#9340) (#34558)
  Handles parsing APK manifests with additional namespaces or attributes (#34535)
  add route information to Flutter.Navigation events (#34508)
  Separate web and io implementations of network image (#34112)
  pass .packages path to snapshot invocation (#34517)
  [Material] Fix TextDirection and selected thumb for RangeSliderThumbShape and RangeSliderValueIndicatorShape (#34501)
  redux of a change to use new engine APIs for Flutter.Frame events (#34521)
  Roll engine b0757e6d308e..f44b7b56623d (4 commits) (#34533)
  Roll engine 2589785b5c5e..b0757e6d308e (6 commits) (#34522)
  Revert "Revert "redux of a change to use new engine APIs for Flutter.Frame events (#34365)" (#34514)" (#34530)
  Make sure fab semantics end up on top (#34512)
  Revert "redux of a change to use new engine APIs for Flutter.Frame events (#34365)" (#34514)
  redux of a change to use new engine APIs for Flutter.Frame events (#34365)

tango5614 commented:
@Piinks You are welcome.I really appreciate your patience.

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -637,12 +647,19 @@ class PageView extends StatefulWidget {
 
 class _PageViewState extends State<PageView> {
-  int _lastReportedPage = 0;
+  /// A flag to record last index reported by [PageView.onPageChanged] and to decide
```
nit: A flag to record the last index...

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -637,12 +647,19 @@ class PageView extends StatefulWidget {
 class _PageViewState extends State<PageView> {
-  int _lastReportedPage = 0;
+  /// A flag to record last index reported by [PageView.onPageChanged] and to decide
+  /// whether to call [PageView.onPageChanged];
```
nit: end with `.` instead of `;`

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -637,12 +647,19 @@ class PageView extends StatefulWidget {
+  /// whether to call [PageView.onPageChanged];
+  int _lastOnPageChangedIndex = 0;
+
+  /// A flag to record last index reported by [PageView.pageDidChange] and to decide
```
the last index

and end with `.`

goderbauer commented on packages/flutter/test/widgets/page_view_test.dart:
```diff
@@ -327,16 +327,44 @@ void main() {
+    expect(log, isEmpty);
+
+    final TestGesture gesture =
+    await tester.startGesture(const Offset(100.0, 100.0));
```
nit: remove the linebreak and put this line right after the `=` above.

goderbauer reviewed (commented):
Sorry for the delay. Just some nits.

goderbauer commented on packages/flutter/test/widgets/page_view_test.dart:
```diff
@@ -480,7 +513,8 @@ void main() {
     await tester.pumpWidget(build(pageSnapping: true));
```
I believe we should make sure that pageDidChange does not trigger before turning snapping back on. Can you add an expect for that here?

tango5614 replied:
Sorry for my mistakes.Actually it did trigger.It was a bug and I have fixed it in the latest commit.

tango5614 committed:
Merge commit 'a0c47e2216a2b50b70c478001cf5652f31a783af' into fix28096

* commit 'a0c47e2216a2b50b70c478001cf5652f31a783af': (187 commits)
  Do not use ideographic baseline for RenderPargraph baseline (#35493)
  Add type to StreamChannel in generated test code. (#35367)
  Fix RenderFittedBox when child.size.isEmpty (#35487)
  add APK build time benchmarks (#35481)
  fix Selection handles position is off (#34665)
  Move usage flutter create tests into memory filesystem. (#35160)
  Include tags in SemanticsNode debug properties (#35491)
  Re-apply 'Add currentSystemFrameTimeStamp to SchedulerBinding' (#35492)
  CupertinoTextField vertical alignment (#34723)
  Mark update-packages as non-experimental (#35467)
  fix default artifacts to exclude ios and android (#35303)
  Roll engine ffba2f68d4e8..7d3e722349e7 (59 commits) (#35489)
  Update macrobenchmarks README and app name (#35477)
  mark windows and macos chrome dev mode as flaky (#35495)
  Use the new service protocol message names (#35482)
  Manual roll of engine 45b66b7...ffba2f6 (#35464)
  more ui-as-code (#35393)
  update reassemble doc (#35164)
  New parameter for RawGestureDetector to customize semantics mapping (#33936)
  Add --target support for Windows and Linux (#34660)
  ...

tango5614 committed:
Modify how to get the value of the currentPage in [PageView.pageDidChange], so that it won't change until the page is totally turned.
Fix some grammatical mistakes.

goderbauer commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -705,7 +705,7 @@ class _PageViewState extends State<PageView> {
         } else if (notification.depth == 0 && widget.pageDidChange != null && notification is ScrollEndNotification) {
           final PageMetrics metrics = notification.metrics;
-          final int currentPage = metrics.page.round();
+          final int currentPage = metrics.page.toInt();
```
This seems odd now. Due to how doubles work, `metrics.page` may be stuck at say 2.999999 even though the transition has completed. In that case we would no longer get a notification that we've reached page three?

tango5614 replied:
You are right.I will revert this line and disable this feature when snapping is off, so `pageDidChange` would never trigger when `pageSnapping`'s value is false.It's that OK?

tango5614 committed:
Disable pageDidChange when pagesnapping is false.

tango5614 commented:
Sorry for the mess. I accidentally rebase my branch.😟

chunhtai commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -682,10 +699,20 @@ class _PageViewState extends State<PageView> {
+            return false;
+          }
+          final PageMetrics metrics = notification.metrics;
+          final int currentPage = metrics.page.round();
```
If user drag the page in the middle pass the page change threshold and release and immediately tap and hold the page, this will trigger the callback while the scroll offset is still in the middle of two pages. Is this a intended behavior?

tango5614 replied:
No, it supposes to trigger exactly when the page has completely changed. Well, seems there is a dilemma.If we use `metrics.page.toInt()`, it sometimes may not trigger because of the double precision problem.
What about this: 
```
final int currentPage = int.parse(metrics.page.toStringAsFixed(15));
```
 Dart doubles are 64-bit floating-point numbers as specified in the IEEE 754 standard.So it has 52 bits of significand, 11 bits of exponent, and 1 sign bit. So the error's value is about (1/2) ^ -52, which makes 15 a safe value as fraction digits.There is a discussion about this.[https://stackoverflow.com/questions/1458633/how-to-deal-with-floating-point-number-precision-in-javascript](url).

If this solution is acceptable, there would be no need to close this function when pageSnapping equal to false because it would have no conflict with `onPageChanged`.So the question below could also be solved.

chunhtai commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -682,10 +699,20 @@ class _PageViewState extends State<PageView> {
             widget.onPageChanged(currentPage);
           }
+        } else if (notification.depth == 0 && widget.pageDidChange != null && notification is ScrollEndNotification) {
+          if (!widget.pageSnapping) {
```
Could this apply to pageSnapping equal to false? if not, it will be more clear to mentions this in the doc of pageDidChange

tango5614 committed:
Fix double number precision problem.Reenable pageDidChange when pageSnapping is off.

tango5614 committed:
Merge branch 'master' into fix28096

tango5614 committed:
change `toStringAsFixed()` to `toStringAsPrecision()`.

chunhtai commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -682,10 +699,22 @@ class _PageViewState extends State<PageView> {
+            _lastOnPageChangedIndex = currentPage;
             widget.onPageChanged(currentPage);
           }
+        } else if (notification.depth == 0 && widget.pageDidChange != null && notification is ScrollEndNotification) {
```
if pagesnapping is false, this will only trigger when user release the touch, that mean it will not trigger when user scroll pass the page and still holding down the finger. Based on the documentation of this callback, this might be correct because it mentioned this callback should only be called after scroll event ends. However, it does not fit the use case that you mentioned in https://github.com/flutter/flutter/issues/28096

If you want to disable scrolling after user scroll to certain page, using this call back will likely end up stuck in between pages if pagesnapping is false.

tango5614 replied:
Yes you are right, that's why I added the [WIP] label in title.
Just finished it.That's we I have done: Just close it when `pageSnapping is false`.
.Add a new property -`isBeingDragged` to `ScrollController` to indicate whether user is dragged the scrollView, and add related test case.

Now it works this way:
- If user release finger and then quickly hold the page, it won't trigger.
- If `pageSnapping` is false, never triggers.
- Only triggers when a user's finger is up and the page change movement has ended.

chunhtai replied:
I think we can ignore isBeingDragged and just check if (!widget.pageSnapping).

I might be missing something, but the only case i can think of isBeingDragged is true and currentPage != _lastPageDidChangeIndex and notification is ScrollEndNotification is when user immediately release and hold, and then time it at the right moment to drag to the perfect offset to the next page. This is an extreme corner case, and even then I think it is ok to invoke the callback. It does not worth to expose a new api in ScrollController to handle this case.

chunhtai replied:
It should still pass the test 'PageDidChange does not trigger when user is dragging' even if without isBeingDragged because ScrollEndNotification should only happened after user release his/her finger.

tango5614 replied:
Oh that’s right.So should it be kept?I think I still need add some code to it to make sure user finger’s quickly up and down works properly.

tango5614 replied:
But just image you are now in page 3, if you drag the PageView back to page 2 a little bit, release finger and hold.Then pageDidChange will be called...maybe this case is not that corner.But indeed,api should not be added too casually.Anyway I will work around it tomorrow. It ‘s really late in my time😂

chunhtai replied:
how about we listen on both ScrollEndNotification and ScrollUpdateNotification and invoke the call back whenever
abs(metrics.page - metrics.page.round()) < 10^-13? and currentPage != _lastPageDidChangeIndex, will this cause any unexpected behavior? In this case we can apply to  pageSnapping = false.

Does this call back has to be called at the end of animation?

tango5614 committed:
Add ScrollController a property called `isBeingDragged` to detect whether scrollView is dragged by user.
Disable `pageDidChange` in PageView when pageSnapping is false.
Add related test cases.

dnfield commented:
@tango5614 - based on the feedback here, it seems like it may be challenging to properly cover all the cases with a change to the framework for this.  There appear to be ways to resolve this without changing the API currently though.

Is there a way we could either capture that work around and still leave resolving the cases to the implementer?  Otherwise, it may be best to close this and document how to achieve this with the current API.

tango5614 commented:
@chunhtai I've made a gif about the case in #28096 
![demo](https://user-images.githubusercontent.com/35095715/61424833-6d1e8680-a947-11e9-83d5-2928634ee28f.gif)

And the code:
```dart
          child: PageView(
            physics: _physics,
            onPageChanged: (int i) {
              if (i == 1) {
                setState(() {
                  _physics = NeverScrollableScrollPhysics();
                });
              }
            },
            children: <Widget>[
              Container(
                color: Colors.red,
              ),
              Container(
                color: Colors.blueAccent,
              ),
            ],
          ),
```
The pageView is stuck because `onPageChanged` is invoked in the middle of two pages.

@dnfield Yes, there is a workaround. Just warp the `PageView` with a `NotificationListener`,listen and handle notifications according to ones needs.In this case, handle the `ScrollEndNotification`.
But one may meet all the drawbacks we met here.We just hand the difficulty to framework users. In fact  I feel it odd to do something in the middle of two pages - in `onPageChanged`.

However if you think a new method is no necessary. I'll close this.

chunhtai commented:
> @chunhtai I've made a gif about the case in #28096
> ![demo](https://user-images.githubusercontent.com/35095715/61424833-6d1e8680-a947-11e9-83d5-2928634ee28f.gif)
> 
> And the code:
> 
> ```dart
>           child: PageView(
>             physics: _physics,
>             onPageChanged: (int i) {
>               if (i == 1) {
>                 setState(() {
>                   _physics = NeverScrollableScrollPhysics();
>                 });
>               }
>             },
>             children: <Widget>[
>               Container(
>                 color: Colors.red,
>               ),
>               Container(
>                 color: Colors.blueAccent,
>               ),
>             ],
>           ),
> ```
> 
> The pageView is stuck because `onPageChanged` is invoked in the middle of two pages.
> 
> @dnfield Yes, there is a workaround. Just warp the `PageView` with a `NotificationListener`,listen and handle notifications according to ones needs.In this case, handle the `ScrollEndNotification`.
> But one may meet all the drawbacks we met here.We just hand the difficulty to framework users. In fact I feel it odd to do something in the middle of two pages - in `onPageChanged`.
> 
> However if you think a new method is no necessary. I'll close this.

I agree the behavior of onPageChanged does not solve all the use cases, and a didPageChanged might be more reasonable to add than onPageChanged, but I soon realized it might not be hard to image we will later on require more and more this type of call back such as pageStartedChanging, or pageScrolledAQuarter.... and so on to solve all kinds of use case, and it might start to complicates the pageview api.

That is why I start to think maybe we should let developer to implement their own, or we can provide a more generic callback something like.. onScroll(double page, ScrollNofitication notification).

@dnfield  what do you think?

goderbauer commented:
It would probably be possible to implement all these kinds of callbacks in an external package that wraps PageView with a NotificationListener. Maybe such an external package would be a better solution to baking this into the framework?

tango5614 commented:
You are right.Closed @goderbauer

olof-dev commented:
Is there now an external package that provides a `didPageChanged` / `pageDidChanged` event? I need to know when the user has actually landed on a new page, whereas `onPageChanged` seems to be called when the PageView is halfway to changing, and even gets called twice if the user scrolls back to get to the original page.
